### PR TITLE
feature: new capabilities contract

### PR DESCRIPTION
**Pull request by aureamunoz**
Here is the first commit with the new capabilities, it's not finished but I would like to share with you and have a first code review. I have to take a look to the doc and the tests&examples in order to be more exhaustive
Related to #452

### CODE REVIEW

**Comment by metacosm** [comment]
I will take a look tomorrow. Thanks!

File: annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonRequiredCapability.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Buildable(builderPackage = "io.dekorate.deps.kubernetes.api.builder")
+@Pojo(name = "RequiredCapabilityConfig", mutable = true, superClass = Configuration.class, relativePath = "../config", withStaticAdapterMethod = false,
+  adapter = @Adapter(suffix = "Adapter", relativePath = "../adapter", withMapAdapterMethod = true))
+@Target({ElementType.CONSTRUCTOR, ElementType.TYPE})
+@Retention(RetentionPolicy.SOURCE)
+public @interface HalkyonRequiredCapability {
```
Comment (metacosm) [inline annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonRequiredCapability.java]: Required capabilities also need parameters.

File: annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonRequiredCapability.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  String name() default "";
+
+  String category() default "";
+
+  String type() default "";
+
+  String boundTo() default "";
```
Comment (metacosm) [inline annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonRequiredCapability.java]: You're missing `autoBindable` as well.

File: annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/handler/ComponentHandler.java [hunk 3]
```diff
@@ -101,31 +103,38 @@ public boolean canHandle(Class<? extends Configuration> type) {
...
-    
+
     if (version != null) {
       resources.decorateCustom(ResourceGroup.NAME, new AddRuntimeVersionToComponentDecorator(version));
     }
     resources.decorateCustom(ResourceGroup.NAME, new DeploymentModeDecorator(config.getDeploymentMode()));
     for (Env env : config.getEnvs()) {
       resources.decorateCustom(ResourceGroup.NAME, new AddEnvToComponentDecorator(env));
     }
+    if (config.getCapabilities().length > 0) {
```
Comment (metacosm) [inline annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/handler/ComponentHandler.java]: capabilities might not be present so `config.getCapabilities` might be `null`

**Comment by iocanel** [comment]
@metacosm: Have your concerns been addressed? Is it ok for merge?

**Comment by metacosm** [comment]
@iocanel I need to check to see if it generates the appropriate descriptors.

**Comment by metacosm** [comment]
@aureamunoz what's the status of this PR?

**Comment by aureamunoz** [comment]
You wanted to do some check, right? Do I merge it?

**Comment by iocanel** [comment]
@aureamunoz: After playing a little bit with the pull request I changed my review to `requires changes`. The main reason, is that I hit some issues, possibly dekorate internal and not directly related to this pull request.

However, to be able to further troubleshoot, I would like to have some test cases and also to simplify the annotation structure (as it might have to do with the issues).

**Comment by metacosm** [comment]
I am working on some fixes at least to more properly reflect the new capabilities structure and which I had missed in my review.

**Comment by aureamunoz** [comment]
Hi guys, yes, there are no enough tests, I'm working on that and think about the capabilities. Thank you

**Comment by aureamunoz** [comment]
yes! I'm refactoring this and do the same.

**Comment by aureamunoz** [comment]
I did a refactor and added the capabilities fields to the `HalkyonComponent` but respecting the output structure in the manifests. Also add some tests, wdyt @iocanel @metacosm ?

File: annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonCapabilities.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Buildable(builderPackage = "io.dekorate.deps.kubernetes.api.builder")
+@Pojo(name = "CapabilitiesConfig", mutable = true, superClass = Configuration.class, relativePath = "../config", withStaticAdapterMethod = false,
+  adapter = @Adapter(suffix = "Adapter", relativePath = "../adapter", withMapAdapterMethod = true))
+@Target({ElementType.CONSTRUCTOR, ElementType.TYPE})
+@Retention(RetentionPolicy.SOURCE)
+public @interface HalkyonCapabilities {
```
Comment (metacosm) [inline annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonCapabilities.java]: Is this annotation still needed if you put its element on HalkyonComponent?

File: annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/model/RequiredComponentCapability.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package io.dekorate.halkyon.model;
+
+import io.sundr.builder.annotations.Buildable;
+import io.sundr.builder.annotations.Inline;
+import io.dekorate.deps.kubernetes.api.model.Doneable;
+
+@Buildable(editableEnabled = false, builderPackage = "io.dekorate.deps.kubernetes.api.builder", inline = @Inline(type = Doneable.class, prefix = "Doneable", value = "done"))
+public class RequiredComponentCapability {
```
Comment (metacosm) [inline annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/model/RequiredComponentCapability.java]: Could this be a sub-class of ComponentCapability?

File: annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonCapabilities.java (hunk 4 above)
Comment (aureamunoz) [reply annotations/halkyon-annotations/src/main/java/io/dekorate/halkyon/annotation/HalkyonCapabilities.java]: Not needed anymor, we can remove it.

**Comment by aureamunoz** [comment]
We would need document it also....in other PR may be?

**Comment by metacosm** [comment]
> We would need document it also....in other PR may be?

I'm not sure what needs to be documented, isn't 

> We would need document it also....in other PR may be?

I'm not sure what to document since the halkyon-related documentation already doesn't cover all halkyon-related features… I mean, we could obviously spend the time documenting everything but I'm not sure it's worth it at this point.

**Comment by metacosm** [comment]
> Overall LGTM.
> 
> I don't really fancy the name `RequiredComponentCapability` and I'd prefer something like `ComponentRequirement` or something. But that's not a blocker for me.

Names could be better, indeed. `ComponentRequirement` wouldn't work, though because we also have `ComponentCapability`, which should probably be renamed `ProvidedComponentCapability` if we were to keep `RequiredComponentCapability`. I'll see tomorrow if I can figure out better names.

**Comment by aureamunoz** [comment]
I just added the new annotation in the config.md and modified for the others

**Comment by iocanel** [comment]
@aureamunoz @metacosm: This is approved. Shall we merge or there are more things to add?

**Comment by aureamunoz** [comment]
Nothing to add from my side. Thanks

**Comment by metacosm** [comment]
We can merge

**Comment by metacosm** [comment]
@iocanel why hasn't this been merged already?

**Comment by iocanel** [comment]
> @iocanel why hasn't this been merged already?

@metacosm: No, reason. Let's merge it now.

FWIW, if a PR is approve feel free to merge it.

**Comment by metacosm** [comment]
> > @iocanel why hasn't this been merged already?
> 
> @metacosm: No, reason. Let's merge it now.
> 
> FWIW, if a PR is approve feel free to merge it.

OK, I didn't feel like it was my place to do so… Thanks!

**Pull request merged**